### ISSUE TITLE
Critical Training Issues: NextTokenPredictor Instantiation  + Dataset Time Inconsistency

### ISSUE THREAD

**Issue by kaku2000** [adept-thu/MomAD#10]
Hello Team! 👋

I've been analyzing the NextTokenPredictor implementation and noticed some potential issues that might be affecting training effectiveness. Would like to share my findings and get your thoughts on these observations.

Summary

Two potential issues that might be affecting NextTokenPredictor training:

NextTokenPredictor is instantiated fresh every forward pass, making learning impossible
Dataset shuffling disrupts temporal order, causing the model to learn from randomly sampled scenarios instead of meaningful temporal sequences

Issue 1: NextTokenPredictor Instantiation Pattern 🔥

Observation
The NextTokenPredictor is currently instantiated inside the forward() method rather than being registered as a module in __init__(), which might be affecting the learning process.

Location: motion_planning_head_roboAD.py:592

# Current implementation: Creates new instance every forward pass
if self.last_plan_query.sum()>0:
    nxttokenpredictor = NextTokenPredictor(256, 128)  # New instance with fresh weights
    enhanced_plan_query = nxttokenpredictor(self.last_plan_query, self.last_planning_classification, enhanced_plan_query)
    # Instance goes out of scope after forward pass

Potential Impact
Limited Learning: Fresh weights created every forward pass

Issue 2: Dataset Temporal Order Question 📊

Observation
The dataset training process uses shuffled sampling, which might affect the temporal relationships that the NextTokenPredictor is designed to learn from.

Location
Sample Token Handling: motion_planning_head_roboAD.py:547-548

# Current temporal state management
prev_sample_token = self.nusc.get('sample', sample0_token)["prev"]
if prev_sample_token == "":
    # Reset temporal state for new scene
    self.last_plan_query = torch.zeros([batch_size, 1, 18, 256]).detach()
    # ... other resets
Dataset Shuffling: Multiple locations show shuffled sampling

# From dataset configuration
shuffler_sampler=dict(type='DistributedGroupSampler')  # Shuffles sample order

Potential Impact
Temporal Relationship: last_plan_query might come from non-sequential samples
Context Consistency: Model learns from shuffled temporal relationships

I hope these observations might be helpful

**Comment by SungwookLE** [comment]
Yes, I also noticed this point. The next_token_predictor class is created at every forward step, which seems to introduce unnecessary noise. @
